{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/rishvikk11/Acute-Leukemia-Biomarker-Analysis/blob/main/researchcode.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YuTgLHUb5c2M"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "from sklearn.model_selection import train_test_split, StratifiedKFold\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "from sklearn.svm import SVC\n",
        "from sklearn.feature_selection import RFE\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.metrics import accuracy_score, classification_report\n",
        "import matplotlib.pyplot as plt\n",
        "from sklearn.decomposition import PCA\n",
        "\n",
        "# Step 1: Load your AML merged data and split patient data into three subsets\n",
        "merged_data = pd.read_csv(\"//Users/rishvikkambhampati/Desktop/AML Clinical and Gene Expression Data.csv\")\n",
        "gene_data = pd.read_csv(\"//Users/rishvikkambhampati/Desktop/AML Gene Expression.csv\")\n",
        "death_classification = merged_data[\"Death Classification\"]\n",
        "patient_classification = merged_data[\"Sample ID\"]\n",
        "# removed_columns = merged_data.drop['Age at Diagnosis',\t'Age At Specimen Acquisition',\t'FLT3-ITD Allelic Ratio\tASXL1',\t'Cancer Type',\t'Cancer Type Detailed',\t'Cause Of Death',\t'CEBPA bi-allelic',\t'Center ID',\t'Cohort',\t'Cumulative Chemo',\t'Unique Regiments',\t'No. Of Cumulative Regiments',\t'Unique Treatment Stages',\t'No. Of Unique Treatment Stages',\t'Unique Treatment Types',\t'No. Of Cumulative Treatments',\t'Current Regimen',\t'Current Stage',\t'Disease Stage At Specimen Collection',\t'Diagnosis At Inclusion',\t'ELN2017',\t'Ethnicity Category',\t'FAB BLAST morphology',\t'FLT3-ITD',\t'De Novo',\t'Relapsed',\t'Transformed',\t'Karyotype',\t'Most Recent Treatment Duration',\t'Most Recent Treatment Type',\t'Mutation Count',\t'Diagnosis non AML, MDS, MPN',\t'NPM1',\t'Oncotree Code',\t'Overall Survival (Months)',\t'Overall Survival Status',\t'Other Cytogenetics',\t'Prior Non-Myeloid',\t'Prior Malignancy Radiation',\t'Type Non-Myeloid',\t'Prior MDS',\t'Prior MDS > Two Months',\t'Prior MDS/MPN',\t'Prior MDS/MPN >Two Months',\t'Prior MPN',\t'Prior MPN > Two Months',\t'Race Category',\t'Response Duration To Induction',\t'Response To Induction',\t'RUNX1',\t'Number of Samples Per Patient',\t'Sex',\t'Diagnosis MDS, MPN',\t'Specific Diagnosis At Inclusion',\t'Specimen Groups',\t'Specimen Type',\t'Surface Antigens Immunohistochemical Stains',\t'Time Of Sample Collection, Relative To Inclusion',\t'TMB (nonsynonymous)',\t'TP53 Pathway',\t'Treatment Type Induction',\t'Used In Manuscript\tVariant Summary',\t'SAMPLE_ID']\n",
        "died_before_two_years = []\n",
        "died_after_two_years = []\n",
        "alive = []\n",
        "for i in range(death_classification.size):\n",
        "    if(death_classification[i] == \"Died Before Two Years\"):\n",
        "        died_before_two_years.append(patient_classification[i])\n",
        "    elif(death_classification[i] == \"Died After Two Years\"):\n",
        "        died_after_two_years.append(patient_classification[i])\n",
        "    else:\n",
        "        alive.append(patient_classification[i])\n",
        "\n",
        "updated_data = gene_data.drop(\"Death Classification\", axis = 'columns')\n",
        "updated_data.pop(\"SAMPLE_ID\")\n",
        "died_before_two_years_genes = updated_data[merged_data[\"Death Classification\"] == \"Died Before Two Years\"]\n",
        "died_after_two_years_genes = updated_data[merged_data[\"Death Classification\"] == \"Died After Two Years\"]\n",
        "alive_genes = updated_data[merged_data[\"Death Classification\"] == \"Alive\"]\n",
        "\n",
        "\n",
        "X_train, X_test, y_train, y_test = train_test_split(updated_data, gene_data[\"Death Classification\"], test_size=0.4, random_state=42)\n",
        "\n",
        "model = SVC(kernel = 'linear')\n",
        "model.fit(X_train, y_train)\n",
        "\n",
        "coefficients = model.coef_[0]\n",
        "\n",
        "gene_coefficients = pd.DataFrame(coefficients, columns=[\"Coefficient\"])\n",
        "list_of_columns = list(updated_data.columns)\n",
        "gene_coefficients = gene_coefficients.assign(Genes = list_of_columns)\n",
        "\n",
        "\n",
        "gene_coefficients = gene_coefficients.reindex(gene_coefficients['Coefficient'].abs().sort_values(ascending=False).index)\n",
        "# model.score(X_train, y_train)\n",
        "pd.set_option('display.max_rows', 500)\n",
        "pd.set_option('display.max_columns', 500)\n",
        "\n",
        "for col in updated_data:\n",
        "    if(updated_data[col].dtype == \"int64\"):\n",
        "        updated_data[col] = updated_data[col].astype(\"float64\")\n",
        "\n",
        "\n",
        "\n",
        "\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import scipy.stats as stats\n",
        "import matplotlib.pyplot as plt\n",
        "from sklearn import linear_model\n",
        "from sklearn.model_selection import train_test_split\n",
        "import seaborn as sns\n",
        "from sklearn.metrics import recall_score, f1_score, precision_score, accuracy_score\n",
        "import io\n",
        "\n",
        "\n",
        "\n",
        "aml_data = pd.read_csv(\"//Users/rishvikkambhampati/Desktop/AML Gene Expression v4.csv\")\n",
        "all_data = pd.read_csv(\"//Users/rishvikkambhampati/Desktop/ALL Gene Expression v3.csv\")\n",
        "all_data.pop(\"SAMPLE_ID\")\n",
        "all_data.pop(\"Death Classification\")\n",
        "\n",
        "for col in aml_data:\n",
        "    if(aml_data[col].dtype == \"int64\"):\n",
        "        aml_data[col] = aml_data[col].astype(\"float64\")\n",
        "\n",
        "for col in all_data:\n",
        "    if(all_data[col].dtype == \"int64\"):\n",
        "        all_data[col] = all_data[col].astype(\"float64\")\n",
        "\n",
        "aml_columns = list(aml_data.columns)\n",
        "all_columns = list(all_data.columns)\n",
        "shared_genes = []\n",
        "\n",
        "for gene in all_columns:\n",
        "    for gene2 in aml_columns:\n",
        "        if(gene2 == gene):\n",
        "            shared_genes.append(gene2)\n",
        "\n",
        "f_statistics = []\n",
        "p_values = []\n",
        "for gene in shared_genes:\n",
        "    aml = aml_data[gene]\n",
        "    all = all_data[gene]\n",
        "    f_statistic, p_value = stats.f_oneway(aml, all)\n",
        "    f_statistics.append(f_statistic)\n",
        "    p_values.append(p_value)\n",
        "\n",
        "anova_frame = pd.DataFrame({\"Genes\": shared_genes, \"F-statistic\": f_statistics, \"p-value\": p_values})\n",
        "updated_anova_frame = anova_frame.sort_values(by=[\"F-statistic\"], ascending=False)\n",
        "\n",
        "frames = [aml_data, all_data]\n",
        "combined_data = pd.concat(frames)\n",
        "\n",
        "shared_genes.remove(\"Diagnosis\")\n",
        "X = shared_genes\n",
        "y = 'Diagnosis'\n",
        "\n",
        "multi_train_df, multi_test_df = train_test_split(combined_data, test_size = 0.2, random_state = 1)\n",
        "\n",
        "X_train = multi_train_df[X]\n",
        "y_train = multi_train_df[y]\n",
        "X_test = multi_test_df[X]\n",
        "y_test = multi_test_df[y]\n",
        "\n",
        "logr = linear_model.LogisticRegression()\n",
        "logr.fit(X_train,y_train)\n",
        "pred = logr.predict(X_test)\n",
        "\n",
        "log_odds = logr.coef_\n",
        "odds = np.exp(log_odds)\n",
        "\n",
        "\n",
        "x = combined_data['IDH1']\n",
        "y = combined_data[\"Diagnosis\"]\n",
        "sns.regplot(x=x, y=y, data = combined_data, logistic = True, ci = None)\n",
        "\n",
        "print(odds)\n",
        "print(accuracy_score(y_test, pred))\n",
        "print(precision_score(y_test, pred))\n",
        "print(recall_score(y_test, pred))\n",
        "print(f1_score(y_test, pred))\n",
        "print(shared_genes)\n",
        "\n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "myenv",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.9.17"
    },
    "orig_nbformat": 4,
    "colab": {
      "provenance": [],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}